lib/DxilDia/DxcPixLiveVariables.cpp

```diff
@@ -29,6 +29,35 @@
 
 #include <unordered_map>
 
+namespace std {
+template <> struct hash<std::pair<llvm::DIScope *, llvm::DIScope *>> {
+  size_t
+  operator()(std::pair<llvm::DIScope *, llvm::DIScope *> const &p) const {
+    const uint64_t f = reinterpret_cast<uint64_t>(p.first);
+    const uint64_t s = reinterpret_cast<uint64_t>(p.second);
+    std::hash<uint64_t> h;
+    return (h(f) ^ h(s));
+  }
+};
+} // namespace std
+
+// If a function is inlined, then the scope of variables within that function
+// will be that scope. Since many such callers may inline the function, we
+// actually need a "scope" that's unique to each "instance" of that function.
+// The caller's scope would be unique, but would have the undesirable
+// side-effect of smushing all of the function's variables together with the
+// caller's variables, at least from the point of view of PIX. The pair of
+// scopes (the function's and the caller's) is both unique to that particular
+// inlining, and distinct from the caller's scope by itself.
+static std::pair<llvm::DIScope *, llvm::DIScope *>
```

Comment:
I'm not sure this could uniquely identify an inlined scope. If `B` calls `A`, and `C` and `D` both call `B`, can this scheme distinguish scopes of the two inlined `A`'s?

Reply:
I think that's the case in the included test? A == DeeperInlinedFunction, B == InlinedFunction and C/D are ClosestHitShader0/1...
Plus, I have only been able to come up with situations in which the deeper-inlined function's llvm::DebugLoc::getInlinedAtScope returns the outermost caller, i.e. the deeper never returns the next-further-up-the-stack inlined function... Would there be any tricks you can think of where I could get deeper inlined scopes than this?

Reply:
Oh I see what's going on. I assumed `getInlinedAtScope` returns the parent caller, but it actually returns the outer-most caller. I bet if you have ClosestHitShader0 call both `InlinedFunction` and `DeeperInlinedFunction`, the two inlined `ret` variable would end up with the same scope pair.

Reply:
I think the _pair_ would be unique because the first element (i.e. the inlined function's own scope) would be different. Where this _does_ break is the same inline called from two different places from the top-level function. Urf. Need to think about that one.

Reply:
Sorry I mean the two `ret` from the two `DeeperInlinedFunction`'s. Would they not both have the pair `{ DeeperInlinedFunction, ClosestHitShader2 }`?

I was thinking we might need to keep a whole callstack of scopes to uniquely identify a variable, but that wouldn't work either, because the two instances of inlined `DeeperInlinedFunction` would have the exact same call stack.
```
void DeeperInlinedFunction() {}
void InlinedFunction() {
   DeeperInlinedFunction();
   DeeperInlinedFunction();
}
void ClosestHitShader2() {
   InlinedFunction();
}
```

Maybe the actual identifier needs to be a stack of `DILocation`s instead of `DIScope`s. 🤔

Reply:
We don't have to make each variable unique to make PIX happy. The issue comes from PIX thinking that two instances of the same variable (i.e. a local in an inlined function that has been inlined twice or more) are the same, and so only a single storage area is allocated for that variable in the output UAV of the debug instrumentation pass.
If a function is inlined into two different top-level functions in the same module, then prior to this change, PIX would associate the inlined function's locals with one UAV storage area, with the result in PIX that one invocation's variables will leak into the other function. That was happening because the prior code used just the inlined function's scope to look up the UAV storage. (The actual experience in PIX is that you get the same 
If a function is inlined twice into the same top-level, then it's OK to have the same UAV storage, since any individual thread that executes that inline function will write first one value into the UAV storage, and later write another. PIX is perfectly happy with this, and in fact you could argue that it SHOULD be this way, since the inline really is conceptually reusing memory.
The double inline case (i.e. one function inlined into two different functions that are then also inlined in the top-level function) is, from PIX's point of view, the same as the single-inline case: all we need to do is make the UAV storage unique per upper-level function (and, of course, unique per inlined function).

(Some mostly-irrelevant complexity is omitted from this discussion which I'm happy to supply.)

Reply:
Thanks for the exaplanation! I think I get it now 👍